Implement ATTACH of MergeTree table for s3_plain disk

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `ATTACH` of `MergeTree` table for `s3_plain` disk (plus some fixes for `s3_plain`)

Simple example:

```sql
-- BACKUP writes Ordinary like structure
set allow_deprecated_database_ordinary=1;
create database ordinary engine=Ordinary;

create table ordinary.test_backup_attach engine=MergeTree() order by tuple() as select * from numbers(100);
-- NOTE: name of backup ("backup") is significant.
backup table ordinary.test_backup_attach TO Disk('backup_disk_s3_plain', 'backup');

drop table ordinary.test_backup_attach;
attach table ordinary.test_backup_attach (number UInt64) engine=MergeTree() order by tuple() settings storage_policy='attach_policy_s3_plain';
```

<details>

<summary>storage configuration</summary>

```xml
<clickhouse>
    <storage_configuration>
        <disks>
            <backup_disk_s3_plain>
                <type>s3_plain</type>
                <endpoint>http://minio1:9001/root/data/disks/disk_s3_plain/</endpoint>
                <access_key_id></access_key_id>
                <secret_access_key></secret_access_key>
            </backup_disk_s3_plain>
            <attach_disk_s3_plain>
                <type>s3_plain</type>
                <!-- NOTE: /backup/ is a name of BACKUP -->
                <endpoint>http://minio1:9001/root/data/disks/disk_s3_plain/backup/</endpoint>
                <access_key_id></access_key_id>
                <secret_access_key></secret_access_key>
            </attach_disk_s3_plain>
        </disks>
        <policies>
            <attach_policy_s3_plain>
                <volumes>
                    <main>
                        <disk>attach_disk_s3_plain</disk>
                    </main>
                </volumes>
            </attach_policy_s3_plain>
        </policies>
    </storage_configuration>
    <backups>
        <allowed_disk>backup_disk_s3_plain</allowed_disk>
    </backups>
</clickhouse>
```

</details>

Follow-up for: #42232
Fixes: #42153 (cc @alexey-milovidov )
Refs: #28527 (cc @filimonov )

_P.S. Note, that this PR also includes some refactoring around metadata code, but everything is in a separate patch, hope this will help with review._

azat committed:
Fix path to files in MetadataStorageFromPlainObjectStorage

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement MetadataStorageFromPlainObjectStorage::getLastModified() (used by MergeTree)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement MetadataStorageFromPlainObjectStorage::getLastChanged() (as for web)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add a comment for IObjectStorage::listPrefix()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove listPrefix() implementation for disks that does not support send_metadata

The reason for removing is it because not compatible with restoring
(with send_metadata set) anyway:
- HDFS - is not compatible with send_metadata, and besides it's
  implementaion is not correct, since it is simply `ls -l`, while the
  following is required: `find . -maxdepth 1 -type f`
- Web - is not compatible with send_metadata anyway
- Local - is not compatible with send_metadata anyway

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Introduce IObjectStorage::listPrefixInPath()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement S3ObjectStorage::listPrefixInPath()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Introduce ReadOnlyMetadataStorage

And use it for:
- MetadataStorageFromPlainObjectStorage
- MetadataStorageFromStaticFilesWebServer

This will allow to reduce ~100-200 lines of duplicated code, and plus
make the code less error prone.

Note, for now I tried to make this without behaviour changes.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement ATTACH of MergeTree table for s3_plain disk

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Slightly better MetadataStorageFromPlainObjectStorage::isFile()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Slightly optimize MetadataStorageFromPlainObjectStorage::isDirectory()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: cover ATTACH of BACKUP from s3_plain disk

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Stress test (tsan) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__tsan_.html)

- #42638 

>[Stress test (ubsan) — Cannot start clickhouse-server](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__ubsan_.html)

```
2022.10.25 09:29:44.584985 [ 1864119 ] {} <Error> Application: DB::Exception: Part `all_1_1_0` (`data/test_19/test/all_1_1_0/`) was found on disk `local_cache_3` which is not defined in the storage policy (defined disks: s3, wrapped disks: ): Cannot attach table `test_19`.`test` from metadata file /var/lib/clickhouse/metadata/test_19/test.sql from query ATTACH TABLE test_19.test (`x` Tuple(UInt64, UInt64)) ENGINE = MergeTree ORDER BY x SETTINGS index_granularity = 8192
```

>[Stress test (debug) — Cannot start clickhouse-server](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__debug_.html)

Looks like the culprit is `s3_check_objects_after_upload`

vitlibar commented on src/Disks/ObjectStorages/IObjectStorage.h:
```diff
@@ -66,7 +66,32 @@ class IObjectStorage
+    /// @param path - path to list prefix in
+    /// @param children - list of child nodes
+    /// @param common_prefixes - list of common prefixes in this path, for unix
+    /// filesystem it is simply subfolders.
```
To get list of subfolders you should use `-type d`, not `-type f`

azat replied:
Actually it is even `find . -maxdepth 1`, folders goes to `common_prefixes` and files to `children`

vitlibar commented on src/Disks/ObjectStorages/LocalObjectStorage.cpp:
```diff
@@ -104,13 +104,6 @@ std::unique_ptr<WriteBufferFromFileBase> LocalObjectStorage::writeObject( /// NO
     return std::make_unique<WriteBufferFromFile>(path, buf_size, flags);
 }
 
-void LocalObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
```
Why did you remove that?

azat replied:
It hadn't been compatible with the only user of this API - restore for `send_metadata`, for details see 333ac44c1483548e8e26daf050fcb98bf1cb478f

vitlibar replied:
Can you explain a bit more why it had been incompatible? I mean for `LocalObjectStorage` the implementation of those functions seems like it must be quite simple and standard.

azat replied:
>Can you explain a bit more why it had been incompatible?

```
void LocalObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
{
    fs::directory_iterator end_it;
    for (auto it = fs::directory_iterator(path); it != end_it; ++it)
        children.emplace_back(it->path().filename(), it->file_size());
}
```

As you can see it does not uses `directory_recursive_iterator` here, and proper `listPrefix` should:
a) be recursive
b) return only files

Both of this wasn't satisfied.
Plus they are not used in the code anyway, since you cannot store metadata (`send_metadata`) in the local disk, and here is the piece of code that creates them:

```
DiskObjectStoragePtr DiskLocal::createDiskObjectStorage()
{
    auto object_storage = std::make_shared<LocalObjectStorage>();
    auto metadata_storage = std::make_shared<FakeMetadataStorageFromDisk>(
        /* metadata_storage */std::static_pointer_cast<DiskLocal>(shared_from_this()),
        object_storage,
        /* object_storage_root_path */getPath());

    return std::make_shared<DiskObjectStorage>(
        getName(),
        disk_path,
        "Local",
        metadata_storage,
        object_storage,
        false, // <-- here, this is send_metadata=false
        /* threadpool_size */16
    );
}
```

> I mean for LocalObjectStorage the implementation of those functions seems like it must be quite simple and standard.

Yes, it can, but I prefer to remove them to avoid ambiguity, anyway this code cannot be tested now.

vitlibar replied:
Yeah, I got it, so it was not incompatible, it was rather wrong.

I'd implement those functions for LocalObjectStorage anyway even if they probably won't be used right now. Just in the sake of consistency and kind of an alternative way to document those too functions better (because operations with filesystem can be considered as basic and simplest). Though I don't insist.

azat replied:
My point is that implementing them w/o tests and w/o actual usage in the code is useless and can be even worse, since they can not work correctly and someone may rely on this.
Anyway I think that even if this should be done, this can be done separately.

vitlibar commented on src/Disks/ObjectStorages/IObjectStorage.h:
```diff
@@ -66,7 +66,32 @@ class IObjectStorage
+    /// @param children - list of child nodes
+    /// @param common_prefixes - list of common prefixes in this path, for unix
+    /// filesystem it is simply subfolders.
+    virtual void listPrefixInPath(const std::string & path,
```
Can we name those functions in more straightforward way like `findFiles` & `getDirectories()` ?

azat replied:
I guess, but I did not want to mix files/folders with S3/GCS/Azure/e.t.c. since they do not have such terms.
And also it is not `findAllFiles` but, `findAllFilesAndDirectories`.
I also thought about adding this is an argument, but came to conclusion that it will be worse.

vitlibar replied:
I mean `listPrefix()` can be named `findAllFiles` (it kind of finds all files in a directory recursively) and `listPrefixInPath` can be named `getDirectoryContents()` (it lists both files and subdirectories in a directory).

azat replied:
Ok, will rename. I don't have strong feeling about this namings.

vitlibar commented on src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h:
```diff
@@ -20,7 +21,7 @@ namespace DB
 /// structure as on disk MergeTree, and does not requires metadata from local
 /// disk to restore.
-class MetadataStorageFromPlainObjectStorage final : public IMetadataStorage
+class MetadataStorageFromPlainObjectStorage final : public ReadOnlyMetadataStorage
```
Why is it inherited from ReadOnlyMetadataStorage? I mean `MetadataStorageFromPlainObjectStorage` doesn't look like it should be readonly. Probably a comment is needed.

azat replied:
Yes, it is not completely readonly, but it does not have any extended metadata support (chown/mtime/stat/...) and `ReadOnlyMetadataStorage` throws `NOT_IMPLEMENTED` exception for this API.

I thought about the name for `ReadOnlyMetadataStorage`, and everything else looks more complex and less cleaner.

Maybe you have any other ideas?

>Probably a comment is needed

Ok (at first I had a comment for ReadOnlyMetadataTransaction/ReadOnlyMetadataStorage but eventually decided that it will only confuse reader)

vitlibar commented on tests/integration/test_attach_backup_from_s3_plain/test.py:
```diff
@@ -0,0 +1,40 @@
+    """
+    )
+
+    assert int(node.query("select count() from ordinary.test_backup_attach")) == 100
```
What will happen if ClickHouse tries to merge parts of `test_backup_attach` ?

azat replied:
It will throw some error, since it does not support mtime/move/...

vitlibar replied:
But merges will be assigned automatically, won't they? So it seems after a while such MergeTree table will spam in logs that it cannot do merging.

azat replied:
Yes and this is like with `web` disk right now, eventually you will get some errors in the log, for proper support some "readonly" property of tables should be implemented.

And I completely agree with you that this should be done.

P.S. moreover there are random SIGSEGVs in web disk when you will try to do this and not only this.

azat replied:
So to clarify, I'm suggesting fixing this separately (since there are other places to fix, like you should not do OPTIMIZE for read-only tables, and isReadOnly should be extended)

vitlibar replied:
Yes I agree, the current implementation must be useful, some issues can be fixed later.

vitlibar commented:
What do you thing can we implement something like `hdfs_plain` disk in the future?

azat commented:
>What do you thing can we implement something like hdfs_plain disk in future?

Sure, it will be very simple.

azat committed:
Rename s/listPrefix/findAllFiles, s/listPrefixInPath/getDirectoryContents/

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

vitlibar commented on src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:
```diff
@@ -141,7 +141,7 @@ std::unique_ptr<WriteBufferFromFileBase> AzureObjectStorage::writeObject( /// NO
 }
 
-void AzureObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
+void AzureObjectStorage::findAllFiles(const std::string & path, RelativePathsWithSize & children) const
```
Is this implementation correct?

azat replied:
I think so, but I have not checked.

azat replied:
I leaved it since it can be used for `send_metadata`

vitlibar commented on src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp:
```diff
@@ -101,18 +101,6 @@ std::unique_ptr<WriteBufferFromFileBase> HDFSObjectStorage::writeObject( /// NOL
-
-    for (int32_t i = 0; i < num_entries; ++i)
-        children.emplace_back(files_list[i].mName, files_list[i].mSize);
-}
```
Was this implementation wrong?

azat replied:
Yes, since this is analog of `ls -l` AFAIR, and it is not recursive.

vitlibar replied:
So you've deleted it for now but in order to implement `hdfs_plain` we'll need to add it back, right? In this case I suppose it would be better to keep the old implementation but probable comment it with `#if 0` and a comment.

azat replied:
>So you've deleted it for now but in order to implement hdfs_plain we'll need to add it back, right? In this case I suppose it would be better to keep the old implementation but probable comment it with #if 0 and a comment.

For it we need both, `findAllFiles` and `getDirectoryContents`, so it should be reimplemented anyway.

vitlibar commented on src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:
```diff
@@ -0,0 +1,165 @@
+
+/// Readonly, throws NOT_IMPLEMENTED error.
+/// Can be used to add limited read-only support of MergeTree.
+class ReadOnlyMetadataTransaction : public IMetadataTransaction
```
Probably bad naming. "Read-only transaction" looks strange. I mean a transaction is always changing something in a batch, how can it be readonly?

azat replied:
Yes, I thought about for awhile but did not managed to came up with a better alternative. Do you have some ideas?

P.S. I thought about using Limited/Surrogate/... but all of these also sound strange and does not clarify it a lot.

vitlibar replied:
Maybe `FakeMetadataTransactionForReadOnlyStorage` will be more clear, what do you think?

azat replied:
a) it is toooo long
b) there is already `FakeMetadataStorageFromDisk` 
c) it does not give any benefit since you do not need transactions for readonly storage

Ok, let's just throw NOT_IMPLEMENTED from all methods.

vitlibar replied:
ok

azat commented on src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:
```diff
@@ -0,0 +1,165 @@
+        return {};
+    }
+
+    struct stat stat(const String & /* path */) const override
```
I prefer to throw NOT_IMPLEMENTED for every API that is not required for now, since who knows who can rely on this...

vitlibar replied:
It's perfectly fine to throw NOT_IMPLEMENTED when you don't have an implementation. It just looks confusing because for example `getLastChanged()` returns 0 and `readFileToString()` throws an exception. Why is that difference?

azat replied:
This was based on `MetadataStorageFromStaticFilesWebServer`, and for it `getLastChanged` returns `0`.
But AFAICS MergeTree code relies on it, though for these type of disks (s3_plain and web) it should not work anyway and returning 0 is not good since this breaks the behaviour.

azat replied:
While for example getLastModified required by MergeTree to work, even for read-only.

vitlibar replied:
The main problem with those two classes `ReadOnlyMetadataStorage` and `ReadOnlyMetadataTransaction` is that they're not consistent with themselves or at least with what seems from their names. I'd prefer them to be like this:

```
class ReadOnlyMetadataTransaction
{
   public:
      // all state-changing functions throw exception "that work because object storage is readonly"
      // maybe except commit (because there is always nothing to commit for this kind of transaction
};

class ReadOnlyMetadataStorage
{
   public:
      // all state-changing functions throw exception "that doesn't work because object storage is readonly"

      // state-reading functions are either left pure virtual or (if it's really necessary) throw exception 
      // but with a different text (just "that is not implemented", not like the message we throw for state-changing functions)

      // and if you want some state-reading or state-changing function to be something else
      // it can be done by overriding functions in inherited class
};
```

azat committed:
Throw NOT_IMPLEMENTED form ReadOnlyMetadataStorage::getLastChanged()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make ReadOnlyMetadataStorage really readonly

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

vitlibar commented on src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:
```diff
@@ -0,0 +1,146 @@
+private:
+    [[noreturn]] static void throwNotImplemented()
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Operation is ont implemented");
```
**not** implemented

Could you please also make `getLastModified` and `getHardlinkCount` follow the same pattern (I mean those function should rather call `throwNotImplemented()` because what they do right now has nothing to do with readonlyness.

azat replied:
And now this becomes looks more and more like `NotImplementedMetadataStorage` then `ReadOnlyMetadataStorage`.
So I'm suggesting remove this classes completely and instead use throw over pure-virtual for the interface classes then.

vitlibar replied:
For transaction probably it's better to keep the readyonly class because it throws a different exception (however it's quite a weak reason).

azat committed:
Remove ReadOnlyMetadataStorage

Throw exceptions from IMetadataStorage instead to avoid introducing
extra abstractions.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Failures does not looks related:

>[Stress test (tsan) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/42628/c93262170df0c0fb09ccbb56f5f73b7a41bc58fc/stress_test__tsan_.html)

- #42640 

>[Integration tests (asan) [1/3] — fail: 1, passed: 666, flaky: 5](https://s3.amazonaws.com/clickhouse-test-reports/42628/c93262170df0c0fb09ccbb56f5f73b7a41bc58fc/integration_tests__asan__[1/3].html)

- `test_distributed_ddl/test.py::test_replicated_without_arguments[configs_secure]`

```
2022-10-31 14:28:37 [ 331 ] DEBUG : Retry 20 got exception Client failed! Return code: 57, stderr: Received exception from server (version 22.11.1):
Code: 57. DB::Exception: Received from 172.16.15.6:9000. DB::Exception: There was an error on [ch4:9440]: Code: 57. DB::Exception: Table test_ordinary.rmt already exists. (TABLE_ALREADY_EXISTS) (version 22.11.1.1). Stack trace:
```

azat commented on tests/integration/test_attach_backup_from_s3_plain/test.py:
```diff
@@ -0,0 +1,40 @@
+    backup table ordinary.test_backup_attach TO Disk('backup_disk_s3_plain', 'backup');
+
+    drop table ordinary.test_backup_attach;
+    attach table ordinary.test_backup_attach (number UInt64) engine=MergeTree() order by tuple() settings storage_policy='attach_policy_s3_plain';
```
And just a note, this will not work for the case with multiple parts, due to files deduplication, but you can workaround this with the following code (but note, that it touches the backup itself, so please, be aware and do not use it on real backups):

```shell
# NOTE: here I use gsutil over s3cmd, since I tested with GCP
gsutil cat gs://path/to/backup_name/.backup > .backup
# NOTE: compose is required to overcome the following issue:
# BadRequestException: 400 Rewriting objects created via Multipart Upload is not implemented yet. As a workaround, you can use compose to overwrite the object (by specifying path with the problem as both the source and output of compose) prior to rewrite.                                               
cat .backup | xq -r '.config.contents.file[] | select(.data_file != null) | [.name, .data_file] | @tsv' | awk -vroot=gs://path/to/backup_name/ '{printf("gsutil cp %s %s || { gsutil compose %s && gsutil cp %s %s; }\n", root $2, root $1, root $2, root $2, root $1)}' > /tmp/backup-data_file-restore.sh
time bash -x backup-data_file-restore.sh |& ts | tee backup-data_file-restore.log
```
